jupyterhub/app.py

```diff
@@ -2005,21 +2011,24 @@ def _user_summary(user):
             spawner._check_pending = False
 
         # parallelize checks for running Spawners
+        # run query on extant Server objects
+        # so this is O(running servers) not O(total users)
         check_futures = []
-        for orm_user in db.query(orm.User):
-            user = self.users[orm_user]
-            self.log.debug("Loading state for %s from db", user.name)
-            for name, orm_spawner in user.orm_spawners.items():
-                if orm_spawner.server is not None:
-                    # spawner should be running
-                    # instantiate Spawner wrapper and check if it's still alive
-                    spawner = user.spawners[name]
-                    # signal that check is pending to avoid race conditions
-                    spawner._check_pending = True
-                    f = asyncio.ensure_future(check_spawner(user, name, spawner))
-                    check_futures.append(f)
-
-        TOTAL_USERS.set(len(self.users))
+        for orm_server in db.query(orm.Server):
+            orm_spawners = orm_server.spawner
+            if not orm_spawners:
+                continue
+            orm_spawner = orm_spawners[0]
```

Comment:
Why only look at the first spawner instead of iterating over all of them?

Reply:
Is it because we look at all Servers and a server only has one spawner? In which case why does `orm_server.spawner` return an array?

Reply:
there will always be one or zero because of how these are created. I'm not sure why this backref is a list instead of a singleton. I might need to fix something in the way the backref is created to make it a singleton, and I just forgot to look into this before moving on.

Reply:
Did some reading about backrefs in sqlalchemy, and now this is properly a 1:1 relationship